discovery-provider/src/model_validator.py

```diff
@@ -0,0 +1,136 @@
+import json
+import copy
+import logging  # pylint: disable=C0302
+from jsonschema import Draft7Validator, ValidationError
+logger = logging.getLogger(__name__)
+
+# https://app.quicktype.io/ -- JSON schema generator
+class ModelValidator:
+    '''
+        {key: value} :
+        {
+            model (first letter uppercase): {
+                model_schema: schema,
+                field_schema: {
+                    title: schema,
+                    mood: schema,
+                    ...
+            },
+                fields: [title, mood, ...]
+            }
+        }
+    '''
+    models_to_schema_and_fields_dict = {}
+
+    @classmethod
+    def validate(cls, to_validate, model, field=""):
+        schema = cls.get_schema(field, model)
+        validator = Draft7Validator(schema)
+
+        try:
+            found_invalid_field = False
+            for error in sorted(validator.iter_errors(to_validate), key=str):
+                found_invalid_field = True
+                logger.warning('Error with {0} instance {1}: {2}'.format(model, to_validate, error.message))
+
+            # if any error occurs, raise exception
+            if found_invalid_field:
+                raise ValidationError('Instance {0} is not proper'.format(to_validate))
+        except ValidationError as e:
+            raise e
+
+    @classmethod
+    def get_schema(cls, field, model):
+        # If model is not present in dict, init its schema and its field subschemas
+        if model not in cls.models_to_schema_and_fields_dict:
+            cls.init_model_schemas(model)
+
+        # If field is empty, return the entire model schema
+        if field == "":
+            return cls.models_to_schema_and_fields_dict[model]['model_schema']
+
+        # Else, return the specified field schema
+        return cls.models_to_schema_and_fields_dict[model]['field_schema'][field]
+
+    @classmethod
+    def init_model_schemas(cls, model):
+        # Load in the model schema in /schemas
+        model_filename = './src/schemas/schema_' + model.lower() + '.json'
+        with open(model_filename) as f:
+            schema = json.load(f)
+
+        model_properties = {
+            "model_schema": schema, # schema for the entire model
+            "field_schema": {}, # schema for just a field in model
+            "fields": schema['definitions'][model]['required'] # list of fields in model
+        }
+
+        cls.models_to_schema_and_fields_dict[model] = model_properties
+
+        # Create all the subschemas for each individual field
+        for field in model_properties['fields']:
+            # Create a deep copy of the entire schema
+            field_schema = copy.deepcopy(schema)
```

Comment:
this naming and logic is super confusing. calling this field_schema and copying the entire json schema and then overwriting the json with a subset of itself is pretty unintuitive

Reply:
in hindsight, i wanted there to only be 1 file necessary for for the validation, whether it be a field or an entire model. i thought that doing a dynamic generation of a subschema  would allow for that, and in the future if we ever change the model, we would only need to update one schema file. 

or, another way to do this is to have 2 schemas:
1. a schema that validates an entire model 
2. a schema that validates only a field of a certain model

with two files, we wouldn't need to generate subschemas. in the future, however, we would need to update 2 places if the model ever changes. 

thoughts? @dmanjunath @jowlee

Reply:
discussed offline, generating subschema is okay. renaming the vars for more clarity!